// ng-mocks/dist/lib/common/core.tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NG_INTERCEPTORS = exports.NG_GUARDS = exports.NG_MOCKS_ROOT_PROVIDERS = exports.NG_MOCKS_INTERCEPTORS = exports.NG_MOCKS_GUARDS = exports.NG_MOCKS_OVERRIDES = exports.NG_MOCKS_TOUCHES = exports.NG_MOCKS = void 0;
var core_1 = require("@angular/core");
exports.NG_MOCKS = new core_1.InjectionToken('NG_MOCKS');
exports.NG_MOCKS_TOUCHES = new core_1.InjectionToken('NG_MOCKS_TOUCHES');
exports.NG_MOCKS_OVERRIDES = new core_1.InjectionToken('NG_MOCKS_OVERRIDES');
exports.NG_MOCKS_GUARDS = new core_1.InjectionToken('NG_MOCKS_GUARDS');
exports.NG_MOCKS_INTERCEPTORS = new core_1.InjectionToken('NG_MOCKS_INTERCEPTORS');
exports.NG_MOCKS_ROOT_PROVIDERS = new core_1.InjectionToken('NG_MOCKS_ROOT_PROVIDERS');
/**
 * Use NG_MOCKS_GUARDS instead.
 * Will be removed in v11.
 * @deprecated
 */
exports.NG_GUARDS = exports.NG_MOCKS_GUARDS;
/**
 * Use NG_MOCKS_INTERCEPTORS instead.
 * Will be removed in v11.
 * @deprecated
 */
exports.NG_INTERCEPTORS = exports.NG_MOCKS_INTERCEPTORS;
