// ng-mocks/dist/lib/common/core.helpers.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extendClass = exports.extractDependency = exports.mapEntries = exports.mapValues = exports.mapKeys = exports.flatten = exports.getInjection = exports.getTestBedInjection = void 0;
var testing_1 = require("@angular/core/testing");
var core_reflect_1 = require("./core.reflect");
exports.getTestBedInjection = function (token) {
    var testBed = testing_1.getTestBed();
    try {
        /* istanbul ignore next */
        return testBed.inject ? testBed.inject(token) : testBed.get(token);
    }
    catch (e) {
        return undefined;
    }
};
exports.getInjection = function (token) {
    var testBed = testing_1.getTestBed();
    /* istanbul ignore next */
    return testBed.inject ? testBed.inject(token) : testBed.get(token);
};
exports.flatten = function (values, result) {
    if (result === void 0) { result = []; }
    if (Array.isArray(values)) {
        values.forEach(function (value) { return exports.flatten(value, result); });
    }
    else {
        result.push(values);
    }
    return result;
};
exports.mapKeys = function (set) {
    var result = [];
    set.forEach(function (_, value) { return result.push(value); });
    return result;
};
exports.mapValues = function (set) {
    var result = [];
    set.forEach(function (value) { return result.push(value); });
    return result;
};
exports.mapEntries = function (set) {
    var result = [];
    set.forEach(function (value, key) { return result.push([key, value]); });
    return result;
};
// Accepts an array of dependencies from providers, skips injections flags,
// and adds the providers to the set.
exports.extractDependency = function (deps, set) {
    var e_1, _a, e_2, _b;
    if (!set) {
        return;
    }
    try {
        for (var deps_1 = __values(deps), deps_1_1 = deps_1.next(); !deps_1_1.done; deps_1_1 = deps_1.next()) {
            var dep = deps_1_1.value;
            if (!Array.isArray(dep)) {
                set.add(dep);
                continue;
            }
            try {
                for (var dep_1 = (e_2 = void 0, __values(dep)), dep_1_1 = dep_1.next(); !dep_1_1.done; dep_1_1 = dep_1.next()) {
                    var flag = dep_1_1.value;
                    if (flag && typeof flag === 'object' && flag.ngMetadataName === 'Optional') {
                        continue;
                    }
                    if (flag && typeof flag === 'object' && flag.ngMetadataName === 'SkipSelf') {
                        continue;
                    }
                    if (flag && typeof flag === 'object' && flag.ngMetadataName === 'Self') {
                        continue;
                    }
                    set.add(flag);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (dep_1_1 && !dep_1_1.done && (_b = dep_1.return)) _b.call(dep_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (deps_1_1 && !deps_1_1.done && (_a = deps_1.return)) _a.call(deps_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
};
exports.extendClass = function (base) {
    var child;
    var parent = base;
    // first we try to eval es2015 style and if it fails to use es5 transpilation in the catch block.
    window.ngMocksParent = parent;
    /* istanbul ignore next */
    try {
        // tslint:disable-next-line:no-eval
        eval("\n        class child extends window.ngMocksParent {\n        }\n        window.ngMocksResult = child\n      ");
        child = window.ngMocksResult;
    }
    catch (e) {
        var ClassEs5 = /** @class */ (function (_super) {
            __extends(ClassEs5, _super);
            function ClassEs5() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            return ClassEs5;
        }(parent));
        child = ClassEs5;
    }
    window.ngMocksParent = undefined;
    // the next step is to respect constructor parameters as the parent class.
    child.parameters = core_reflect_1.jitReflector.parameters(parent);
    return child;
};
