// @nicky-lenaers/ngx-scroll-to/esm2015/lib/scroll-to.module.js
import * as ɵngcc0 from '@angular/core';
var ScrollToModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ScrollToDirective } from './scroll-to.directive';
import { ScrollToService } from './scroll-to.service';
/** Scroll To Module */
let ScrollToModule = ScrollToModule_1 = class ScrollToModule {
    /**
     * Guaranteed singletons for provided Services across App.
     *
     * @return          An Angular Module with Providers
     */
    static forRoot() {
        return {
            ngModule: ScrollToModule_1,
            providers: [
                ScrollToService
            ]
        };
    }
};
ScrollToModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ScrollToModule });
ScrollToModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ScrollToModule_Factory(t) { return new (t || ScrollToModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ScrollToModule, { declarations: function () { return [ScrollToDirective]; }, exports: function () { return [ScrollToDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ScrollToModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ScrollToDirective
                ],
                exports: [
                    ScrollToDirective
                ]
            }]
    }], null, null); })();
export { ScrollToModule };
