// angulartics2/esm2015/angulartics2On.js
import { Directive, ElementRef, Input, NgModule, Renderer2, } from '@angular/core';
import { Angulartics2 } from './angulartics2-core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './angulartics2-core';
export class Angulartics2On {
    constructor(elRef, angulartics2, renderer) {
        this.elRef = elRef;
        this.angulartics2 = angulartics2;
        this.renderer = renderer;
        this.angularticsProperties = {};
    }
    ngAfterContentInit() {
        this.renderer.listen(this.elRef.nativeElement, this.angulartics2On || 'click', (event) => this.eventTrack(event));
    }
    eventTrack(event) {
        const action = this.angularticsAction; // || this.inferEventName();
        const properties = Object.assign(Object.assign({}, this.angularticsProperties), { eventType: event.type });
        if (this.angularticsCategory) {
            properties.category = this.angularticsCategory;
        }
        if (this.angularticsLabel) {
            properties.label = this.angularticsLabel;
        }
        if (this.angularticsValue) {
            properties.value = this.angularticsValue;
        }
        this.angulartics2.eventTrack.next({
            action,
            properties,
        });
    }
}
Angulartics2On.ɵfac = function Angulartics2On_Factory(t) { return new (t || Angulartics2On)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Angulartics2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
Angulartics2On.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: Angulartics2On, selectors: [["", "angulartics2On", ""]], inputs: { angularticsProperties: "angularticsProperties", angulartics2On: "angulartics2On", angularticsAction: "angularticsAction", angularticsCategory: "angularticsCategory", angularticsLabel: "angularticsLabel", angularticsValue: "angularticsValue" } });
Angulartics2On.ctorParameters = () => [
    { type: ElementRef },
    { type: Angulartics2 },
    { type: Renderer2 }
];
Angulartics2On.propDecorators = {
    angulartics2On: [{ type: Input, args: ['angulartics2On',] }],
    angularticsAction: [{ type: Input }],
    angularticsCategory: [{ type: Input }],
    angularticsLabel: [{ type: Input }],
    angularticsValue: [{ type: Input }],
    angularticsProperties: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2On, [{
        type: Directive,
        args: [{ selector: '[angulartics2On]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.Angulartics2 }, { type: ɵngcc0.Renderer2 }]; }, { angularticsProperties: [{
            type: Input
        }], angulartics2On: [{
            type: Input,
            args: ['angulartics2On']
        }], angularticsAction: [{
            type: Input
        }], angularticsCategory: [{
            type: Input
        }], angularticsLabel: [{
            type: Input
        }], angularticsValue: [{
            type: Input
        }] }); })();
export class Angulartics2OnModule {
}
Angulartics2OnModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Angulartics2OnModule });
Angulartics2OnModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Angulartics2OnModule_Factory(t) { return new (t || Angulartics2OnModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Angulartics2OnModule, { declarations: [Angulartics2On], exports: [Angulartics2On] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2OnModule, [{
        type: NgModule,
        args: [{
                declarations: [Angulartics2On],
                exports: [Angulartics2On]
            }]
    }], null, null); })();
