// angulartics2/esm2015/angulartics2.module.js
import { NgModule } from '@angular/core';
import { AngularRouterTracking } from './angular-router';
import { Angulartics2 } from './angulartics2-core';
import { ANGULARTICS2_TOKEN } from './angulartics2-token';
import { Angulartics2On, Angulartics2OnModule } from './angulartics2On';
import { RouterlessTracking } from './routerless';
import * as ɵngcc0 from '@angular/core';
export class Angulartics2Module {
    static forRoot(settings = {}) {
        return {
            ngModule: Angulartics2Module,
            providers: [
                { provide: ANGULARTICS2_TOKEN, useValue: { settings } },
                { provide: RouterlessTracking, useClass: AngularRouterTracking },
                Angulartics2,
            ],
        };
    }
}
Angulartics2Module.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Angulartics2Module });
Angulartics2Module.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Angulartics2Module_Factory(t) { return new (t || Angulartics2Module)(); }, imports: [[Angulartics2OnModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Angulartics2Module, { imports: function () { return [Angulartics2OnModule]; }, exports: function () { return [Angulartics2On]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2Module, [{
        type: NgModule,
        args: [{
                imports: [Angulartics2OnModule],
                exports: [Angulartics2On]
            }]
    }], null, null); })();
