// ng-mocks/dist/lib/mock-service/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerMockFunction = void 0;
var helper_create_mock_from_prototype_1 = require("./helper.create-mock-from-prototype");
var helper_extract_methods_from_prototype_1 = require("./helper.extract-methods-from-prototype");
var helper_extract_properties_from_prototype_1 = require("./helper.extract-properties-from-prototype");
var helper_extract_property_descriptor_1 = require("./helper.extract-property-descriptor");
var helper_mock_1 = require("./helper.mock");
var helper_mock_function_1 = require("./helper.mock-function");
var helper_replace_with_mocks_1 = require("./helper.replace-with-mocks");
var helper_resolve_provider_1 = require("./helper.resolve-provider");
var helper_use_factory_1 = require("./helper.use-factory");
/* istanbul ignore next */
var getGlobal = function () { return window || global; };
// We need a single pointer to the object among all environments.
getGlobal().ngMocksMockServiceHelper = getGlobal().ngMocksMockServiceHelper || {
    mockFunction: helper_mock_function_1.default,
    registerMockFunction: function (func) {
        getGlobal().ngMocksMockServiceHelper.mockFunction.customMockFunction = func;
    },
    createMockFromPrototype: helper_create_mock_from_prototype_1.default,
    extractMethodsFromPrototype: helper_extract_methods_from_prototype_1.default,
    extractPropertiesFromPrototype: helper_extract_properties_from_prototype_1.default,
    extractPropertyDescriptor: helper_extract_property_descriptor_1.default,
    mock: helper_mock_1.default,
    replaceWithMocks: helper_replace_with_mocks_1.default,
    resolveProvider: helper_resolve_provider_1.default,
    useFactory: helper_use_factory_1.default,
};
/**
 * DO NOT USE this object outside of the library.
 * It can be changed any time without a notice.
 *
 * @internal
 */
exports.default = (function () { return getGlobal().ngMocksMockServiceHelper; })();
exports.registerMockFunction = getGlobal().ngMocksMockServiceHelper
    .registerMockFunction;
