// @ng-dynamic-forms/core/esm2015/lib/directive/dynamic-template.directive.js
import { Directive, Input, TemplateRef } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export var DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT;
(function (DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT) {
    DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT["Start"] = "START";
    DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT["End"] = "END";
})(DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT || (DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT = {}));
export class DynamicTemplateDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.align = DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.End;
        this.as = null;
    }
}
DynamicTemplateDirective.ɵfac = function DynamicTemplateDirective_Factory(t) { return new (t || DynamicTemplateDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
DynamicTemplateDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DynamicTemplateDirective, selectors: [["ng-template", "modelId", ""], ["ng-template", "modelType", ""]], inputs: { align: "align", as: "as", index: "index", modelId: "modelId", modelType: "modelType" } });
DynamicTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
DynamicTemplateDirective.propDecorators = {
    align: [{ type: Input }],
    as: [{ type: Input }],
    index: [{ type: Input }],
    modelId: [{ type: Input }],
    modelType: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DynamicTemplateDirective, [{
        type: Directive,
        args: [{
                selector: "ng-template[modelId],ng-template[modelType]"
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { align: [{
            type: Input
        }], as: [{
            type: Input
        }], index: [{
            type: Input
        }], modelId: [{
            type: Input
        }], modelType: [{
            type: Input
        }] }); })();
