// @ng-dynamic-forms/core/esm2015/lib/component/dynamic-form-control-with-template.component.js
import { Directive } from "@angular/core";
import { DynamicFormControlComponent } from "./dynamic-form-control.component";
import { isString } from "../utils/core.utils";
import { DynamicFormLayoutService } from "../service/dynamic-form-layout.service";
import { DynamicFormValidationService } from "../service/dynamic-form-validation.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../service/dynamic-form-layout.service';
import * as ɵngcc2 from '../service/dynamic-form-validation.service';
export class DynamicFormControlWithTemplateComponent extends DynamicFormControlComponent {
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
    }
    ngAfterViewInit() {
        this.layoutService
            .filterTemplatesByModel(this.model, this.templates)
            .forEach(template => this.bindTemplate(template));
    }
    bindTemplate(template) {
        if (isString(template.as) && this.templateDirectives.has(template.as)) {
            const property = this.templateDirectives.get(template.as);
            this.viewChild[property] = this.mapTemplate(template);
        }
    }
}
DynamicFormControlWithTemplateComponent.ɵfac = function DynamicFormControlWithTemplateComponent_Factory(t) { return new (t || DynamicFormControlWithTemplateComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DynamicFormLayoutService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DynamicFormValidationService)); };
DynamicFormControlWithTemplateComponent.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DynamicFormControlWithTemplateComponent, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
DynamicFormControlWithTemplateComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DynamicFormControlWithTemplateComponent, [{
        type: Directive
    }], function () { return [{ type: ɵngcc1.DynamicFormLayoutService }, { type: ɵngcc2.DynamicFormValidationService }]; }, null); })();
