// @ng-dynamic-forms/core/esm2015/lib/model/file-upload/dynamic-file-upload.model.js
import { __decorate, __metadata } from "tslib";
import { DynamicFileControlModel } from "../dynamic-file-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isNumber } from "../../utils/core.utils";
export const DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD = "FILE_UPLOAD";
export class DynamicFileUploadModel extends DynamicFileControlModel {
    constructor(config, layout) {
        var _a, _b;
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD;
        this.accept = Array.isArray(config.accept) ? config.accept : null;
        this.autoUpload = isBoolean(config.autoUpload) ? config.autoUpload : true;
        this.maxSize = isNumber(config.maxSize) ? config.maxSize : null;
        this.minSize = isNumber(config.minSize) ? config.minSize : null;
        this.removeUrl = (_a = config.removeUrl) !== null && _a !== void 0 ? _a : null;
        this.showFileList = isBoolean(config.showFileList) ? config.showFileList : true;
        this.url = (_b = config.url) !== null && _b !== void 0 ? _b : null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFileUploadModel.prototype, "accept", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "autoUpload", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "maxSize", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "minSize", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "removeUrl", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "showFileList", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "url", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "type", void 0);
