// @ng-dynamic-forms/core/esm2015/lib/model/dynamic-input-control.model.js
import { __decorate, __metadata } from "tslib";
import { DynamicFormValueControlModel } from "./dynamic-form-value-control.model";
import { serializable } from "../decorator/serializable.decorator";
import { isBoolean, isNumber } from "../utils/core.utils";
export class DynamicInputControlModel extends DynamicFormValueControlModel {
    constructor(config, layout) {
        var _a, _b, _c, _d;
        super(config, layout);
        this.autoComplete = (_a = config.autoComplete) !== null && _a !== void 0 ? _a : "on";
        this.autoFocus = isBoolean(config.autoFocus) ? config.autoFocus : false;
        this.maxLength = isNumber(config.maxLength) ? config.maxLength : null;
        this.minLength = isNumber(config.minLength) ? config.minLength : null;
        this.placeholder = (_b = config.placeholder) !== null && _b !== void 0 ? _b : "";
        this.prefix = (_c = config.prefix) !== null && _c !== void 0 ? _c : null;
        this.readOnly = isBoolean(config.readOnly) ? config.readOnly : false;
        this.spellCheck = isBoolean(config.spellCheck) ? config.spellCheck : false;
        this.suffix = (_d = config.suffix) !== null && _d !== void 0 ? _d : null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "autoComplete", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "autoFocus", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicInputControlModel.prototype, "maxLength", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicInputControlModel.prototype, "minLength", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "placeholder", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "prefix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "readOnly", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "spellCheck", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "suffix", void 0);
