// @ng-dynamic-forms/core/esm2015/lib/model/dynamic-form-value-control.model.js
import { __decorate, __metadata } from "tslib";
import { BehaviorSubject } from "rxjs";
import { DynamicFormControlModel } from "./dynamic-form-control.model";
import { serializable } from "../decorator/serializable.decorator";
import { isBoolean, isObject } from "../utils/core.utils";
export class DynamicFormValueControlModel extends DynamicFormControlModel {
    constructor(config, layout) {
        var _a, _b, _c;
        super(config, layout);
        this.additional = isObject(config.additional) ? config.additional : null;
        this.hint = (_a = config.hint) !== null && _a !== void 0 ? _a : null;
        this.required = isBoolean(config.required) ? config.required : false;
        this.tabIndex = (_b = config.tabIndex) !== null && _b !== void 0 ? _b : null;
        this.value$ = new BehaviorSubject((_c = config.value) !== null && _c !== void 0 ? _c : null);
        this.value$.subscribe(value => this._value = value);
        this.valueChanges = this.value$.asObservable();
    }
    get value() {
        return this.value$.getValue();
    }
    set value(value) {
        this.value$.next(value);
    }
    getAdditional(key, defaultValue) {
        return this.additional !== null && this.additional.hasOwnProperty(key) ? this.additional[key] : defaultValue;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "additional", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormValueControlModel.prototype, "hint", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFormValueControlModel.prototype, "required", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFormValueControlModel.prototype, "tabIndex", void 0);
__decorate([
    serializable("value"),
    __metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "_value", void 0);
