// angulartics2/routerlessmodule/esm2015/routerless.module.js
import { NgModule } from '@angular/core';
import { Angulartics2, Angulartics2OnModule, ANGULARTICS2_TOKEN, RouterlessTracking, } from 'angulartics2';
import * as ɵngcc0 from '@angular/core';
export class Angulartics2RouterlessModule {
    static forRoot(settings = {}) {
        return {
            ngModule: Angulartics2RouterlessModule,
            providers: [
                { provide: ANGULARTICS2_TOKEN, useValue: { settings } },
                RouterlessTracking,
                Angulartics2,
            ],
        };
    }
}
Angulartics2RouterlessModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Angulartics2RouterlessModule });
Angulartics2RouterlessModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Angulartics2RouterlessModule_Factory(t) { return new (t || Angulartics2RouterlessModule)(); }, imports: [[Angulartics2OnModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Angulartics2RouterlessModule, { imports: function () { return [Angulartics2OnModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2RouterlessModule, [{
        type: NgModule,
        args: [{
                imports: [Angulartics2OnModule]
            }]
    }], null, null); })();
