// angulartics2/routerlessmodule/bundles/angulartics2-routerlessmodule.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('angulartics2'),exports, require('@angular/core'), require('angulartics2')) :
    typeof define === 'function' && define.amd ? define('angulartics2/routerlessmodule', ['@angular/core','angulartics2','exports', '@angular/core', 'angulartics2'], factory) :
    (global = global || self, factory(global.ng.core,global.angulartics2,(global.angulartics2 = global.angulartics2 || {}, global.angulartics2.routerlessmodule = {}), global.ng.core, global.angulartics2));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, angulartics2) { 'use strict';

    var Angulartics2RouterlessModule = /** @class */ (function () {
        function Angulartics2RouterlessModule() {
        }
        Angulartics2RouterlessModule.forRoot = function (settings) {
            if (settings === void 0) { settings = {}; }
            return {
                ngModule: Angulartics2RouterlessModule,
                providers: [
                    { provide: angulartics2.ANGULARTICS2_TOKEN, useValue: { settings: settings } },
                    angulartics2.RouterlessTracking,
                    angulartics2.Angulartics2,
                ],
            };
        };
Angulartics2RouterlessModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Angulartics2RouterlessModule });
Angulartics2RouterlessModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Angulartics2RouterlessModule_Factory(t) { return new (t || Angulartics2RouterlessModule)(); }, imports: [[angulartics2.Angulartics2OnModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Angulartics2RouterlessModule, { imports: [ɵngcc1.Angulartics2OnModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Angulartics2RouterlessModule, [{
        type: core.NgModule,
        args: [{
                imports: [angulartics2.Angulartics2OnModule]
            }]
    }], function () { return []; }, null); })();
        return Angulartics2RouterlessModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.Angulartics2RouterlessModule = Angulartics2RouterlessModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
